### PR TITLE
#167512374 User can get article category subscribed for

### PR DESCRIPTION
**Pull request by Jonathan4github**
#### What does this PR do?
This Pull request implements user should be able to get articles category subscribed for
#### Description of Task to be completed
- create a subscription model
- implement functionality in the controller file
#### How should this be manually tested?
 - clone this repository
- ```cd``` into the repo and run ```npm install```
- using postman **SET**  ```Authorization``` to ```Bearer <user token>```
- make a **GET**  request to /api/v1/subscriptions 
#### What are the relevant pivotal tracker stories?

[#167512374](https://pivotaltracker.com/story/show/167512374)

#### Screenshots (if appropriate)
Screenshot 2019-08-19 at 9 07 23 PM

### CODE REVIEW

File: tests/routes/articles.test.js [hunk 1]
```diff
@@ -883,3 +895,48 @@ describe('GET /articles/articleId/inline_comments', () => {
...
+  it('should return 404 Resource not found', async () => {
+    const { status } = await chai
+      .request(app)
+      .get(`${baseRoute}/articles/subscriptions`)
+      .set('Authorization', `Bearer ${userToken}`);
+
+    status.should.eql(404);
+  });
+
+  it('should return 200', async () => {
```
Comment (Jonathan4github) [inline tests/routes/articles.test.js]: Add message for 404

File: src/controllers/article.js [hunk 2]
```diff
@@ -516,5 +521,62 @@ export default {
...
+  /** controller for getting articles subscribed for.
+   *
+   * @function
+   *
+   * @param {Object} request - express request object
+   * @param {Object} response - express response object
+   *
+   * @return {Object} - callback that execute the controller
+   */
+  getArticlesSubscribed: async (request, response) => {
```
Comment (Jonathan4github) [inline src/controllers/article.js]: getSubscribedArticles

File: src/controllers/article.js [hunk 3]
```diff
@@ -516,5 +521,62 @@ export default {
...
+    const { id: userId } = request.user;
+    const { page = 1, limit = 10 } = request.query;
+
+    const user = await Users.findByPk(userId);
```
Comment (Jonathan4github) [inline src/controllers/article.js:537]: Move to verify token middleware.

File: src/controllers/article.js [hunk 4]
```diff
@@ -516,5 +521,62 @@ export default {
...
+
+    if (!user) throw new NotFoundError();
+
+    const categories = (await user.getSubscriptions({ attributes: ['categoryId'] })).map(sub => sub.categoryId);
```
Comment (Jonathan4github) [inline src/controllers/article.js]: categoryIds

File: src/controllers/article.js [hunk 5]
```diff
@@ -516,5 +521,62 @@ export default {
...
+        attributes: [['categoryId', 'id']],
+        include: [{
+          model: Categories,
+          as: 'tag',
+          attributes: ['category']
+        }]
+      }]
+    };
+
+    const { data, ...otherDatas } = await paginator(Articles, query);
```
Comment (Jonathan4github) [inline src/controllers/article.js]: other data should be singular

File: src/controllers/article.js [hunk 6]
```diff
@@ -516,5 +521,62 @@ export default {
...
+        ...articleData,
+        category: category.map(({ id, tag }) => ({ id, category: tag && tag.category }))
+      };
+    });
+
+    return response.status(200).json({
+      status: 'success',
+      data: articles,
+      ...otherDatas,
+      message: 'Article category subscribed for retrieved successfully'
```
Comment (Jonathan4github) [inline src/controllers/article.js]:
```suggestion
      message: 'Article categories subscribed for retrieved successfully'
```

File: src/routes/article.js [hunk 7]
```diff
@@ -296,6 +297,8 @@ router.post(
   asyncWrapper(createArticle)
 );
 
+router.get('/subscriptions', asyncWrapper(verifyToken), asyncWrapper(getArticlesSubscribed));
```
Comment (Jonathan4github) [inline src/routes/article.js]: Swagger doc

File: tests/routes/articles.test.js [hunk 8]
```diff
@@ -883,3 +895,48 @@ describe('GET /articles/articleId/inline_comments', () => {
...
+    await ArticleCategories.bulkCreate(subscriptionCategories, { returning: true });
+    await Subscriptions.bulkCreate(subscriptions, { returning: true });
+  });
+
+  after(async () => {
+    await Users.destroy({ where: {} });
+    await Articles.destroy({ where: {} });
+    await ArticleCategories.destroy({ where: {} });
+    await Subscriptions.destroy({ where: {} });
+  });
```
Comment (Jonathan4github) [inline tests/routes/articles.test.js]: newline and make check what test is to return.

File: tests/routes/profile.test.js [hunk 9]
```diff
@@ -30,11 +30,14 @@ let currentUserId;
 
 describe('Profile', () => {
   before(async () => {
-    await Users.destroy({ cascade: true, truncate: true });
+    await Users.destroy({ where: {} });
     responseToken = await chai.request(app)
       .post(signUrl)
       .send(usersignUpdetail);
   });
+  after(async () => {
```
Comment (iverenshaguy) [inline tests/routes/profile.test.js:39]: Leave a new line above and below this `after` block

File: tests/routes/articles.test.js [hunk 10]
```diff
@@ -89,13 +96,20 @@ describe('Articles API', () => {
     it('should throw error if user not found', async () => {
       const response = await chai.request(app)
         .get(`${baseRoute}/articles/feed`)
-        .set('Authorization', `Bearer ${userToken}`);
+        .set('Authorization', `Bearer ${randomUserToken}`);
```
Comment (iverenshaguy) [inline tests/routes/articles.test.js:100]: Why was this changed? Why should a user be random?

File: src/middlewares/authentication.js [hunk 11]
```diff
@@ -33,11 +35,19 @@ export default {
...
 
-    jwt.verify(token, SECRET_KEY, (error, decodedToken) => {
-      if (error) throw new ApplicationError(401, `${error.message}`);
+    jwt.verify(token, SECRET_KEY, async (error, decodedToken) => {
+      if (error) return next(new ApplicationError(401, `${error.message}`));
       request.user = decodedToken;
 
-      next();
+      const { id: userId } = request.user;
+
```
Comment (iverenshaguy) [inline src/middlewares/authentication.js:40]: Please remove this new line

File: src/middlewares/authentication.js [hunk 12]
```diff
@@ -33,11 +35,19 @@ export default {
 
     if (blacklistedToken) throw new ApplicationError(403, 'Please login or signup to access this resource');
 
-    jwt.verify(token, SECRET_KEY, (error, decodedToken) => {
-      if (error) throw new ApplicationError(401, `${error.message}`);
+    jwt.verify(token, SECRET_KEY, async (error, decodedToken) => {
+      if (error) return next(new ApplicationError(401, `${error.message}`));
```
Comment (iverenshaguy) [inline src/middlewares/authentication.js:39]: Why was `throw` changed to return? What benefit does this give us? I know express automatically uses the error middleware when `next()` is called with an argument, what benefit does this give us over simply throwing the error which does the same thing?

File: src/middlewares/authentication.js [hunk 13]
```diff
@@ -33,11 +35,19 @@ export default {
...
       request.user = decodedToken;
 
-      next();
+      const { id: userId } = request.user;
+
+      const user = await Users.findByPk(userId);
+
+      if (!user) return next(new ApplicationError(403, 'User not found'));
```
Comment (iverenshaguy) [inline src/middlewares/authentication.js]: Must we return next? Why not just throw it? Any benefit?

File: src/controllers/article.js (hunk 4 above)
Comment (iverenshaguy) [reply src/controllers/article.js]: Why wasn't this changed?

**Review by iverenshaguy — CHANGES_REQUESTED** [review]
Please review my feedback

File: tests/routes/articles.test.js (hunk 10 above)
Comment (Jonathan4github) [reply tests/routes/articles.test.js:100]: When I added a check if the user exists in the authentication middleware. Most tests failed because the "userToken" does not exist in the database so I changed them to randomUserToken.

**Review by iverenshaguy — APPROVED** [review]
